source/develop/release-management/features/virt/tpm-device.md

```diff
@@ -0,0 +1,154 @@
+
+#### Storing the data in domain XML metadata
+
+Using domain XML metadata, it's possible to transfer TPM data in all the directions.  TPM data can be read from the data on VM start, with or without Engine assistance, updates can be stored there atomically and Engine can update the data in monitoring.
+
+Ensuring data consistency read from swtpm is the same as with API calls.
+
+The only problem with using metadata is that it increases domain XML size significantly.  TPM data shouldn't be large although to know the size limits would require further investigation.  But initial swtpm data for a VM takes about 4 KB encoded.  It would "pollute" the domain XML and would have to be transferred each time Engine asks for the domain XML, even when there are no changes to data, which means it would be waste most of the time.
+
+This approach is a bit simpler and more flexible than using API calls but has the data size drawback.
+
+#### Using shared storage
+
+A somewhat similar approach to using domain XML metadata would be to store the data to a shared storage.  Instead of writing to or reading it from the metadata, it would be written to or read from the storage.  The benefit would be avoiding domain XML pollution and unnecessary data transfers.  Additionally, Engine wouldn't have to process the data and maintain it in the database and the data could be updated even when Engine was temporarily unavailable.
+
+However, there are also drawbacks:
+
+- A volume for TPM data must be created, arranged and maintained properly.
+- Vdsm is responsible for arranging atomic data updates, to avoid ending up with partially written data in case storage gets disconnected.
+- Unless direct I/O is used, Vdsm can't be sure the data was actually written.
+
+This approach would be more complicated than using metadata but it would have the advantages mentioned above.
+
+#### Making swtpm to use shared storage directly
+
+The swtpm data location is selected by libvirt automatically but we could probably make a symbolic link to the desired shared location.  This approach would avoid data handling on the Vdsm side and all the updates would be stored immediately.  But it shares some of the drawbacks of using shared storage and there are additional risks:
+
+- Data consistency is at risk when storage gets disconnected during swtpm write.
+- swtpm reacts to write failures [by deleting the data completely](https://github.com/stefanberger/swtpm/blob/609dfd873a31fc0402752c50d1d5cfa58949507f/src/swtpm/swtpm_nvfile.c#L563), which may fail as well, but a question is if the data is ever attempted to write again without further updates.
+- We can't depend on all the swtpm implementation details.
+- libvirt removes the data after domain gets undefined.  We would have to prevent that, perhaps by removing the symbolic link before undefining the libvirt domain.
```

Comment:
We can ask libivrt to support external tpm data, so libvirt does not create or remove anything. This way we have
complete control over the data, and we don't need to change libvirt flows.

So we can do:

1. Store tpm data on the host tpm store
2. Start vm with tpm store path
3. Stop vm (undefine, etc.)
4. Pass current tpm data to engine

We use the same method for other stuff like scratch disks during backup. Libvirt supports EXTERNAL flags to let the user control something that libvirt cannot mange. This sounds like the best approach.

Reply:
This would be necessary only when swtpm was writing to the shared storage directly. I like that option the least of all. In the other cases it makes sense that libvirt removes related data when we undefine a domain -- it's up to us to delay cleanup if we need that.